# 01. Importing Libraries

In [113]:
# Import Libraries
import pandas as pd
import numpy as np
import os

# 02. Importing Data

In [ ]:
# Importing 'products.csv' as as df_prods from 'Original Data' folder and 'order.cvs' 
# as df_ords from 'Prepared Data' folder

In [114]:
path_prods = r'/Users/robertochidiac/Desktop/Instacart Basket Analysis/Data/Original Data/products.csv'

In [115]:
path_prods

'/Users/robertochidiac/Desktop/Instacart Basket Analysis/Data/Original Data/products.csv'

In [116]:
path_ords = r'/Users/robertochidiac/Desktop/Instacart Basket Analysis/Data/Prepared Data/orders_wrangled.csv'

In [117]:
path_ords

'/Users/robertochidiac/Desktop/Instacart Basket Analysis/Data/Prepared Data/orders_wrangled.csv'

In [118]:
df_prods = pd.read_csv(os.path.join(path_prods), index_col = False)

In [119]:
df_ords = pd.read_csv(os.path.join(path_ords), index_col = False)

# 03. Data Consistency Checks

In [120]:
# cerating a test dataframe to practice mixed-type data
# Create a dataframe
df_test = pd.DataFrame()

In [121]:
# Create a mixed-type column
df_test['mix'] = ['a', 'b', 1, True]

In [122]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [123]:
# the function below will check wether the dataframe contains any mix-type colums
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [124]:
# the code below will change the data type
df_test['mix'] = df_test['mix'].astype('str')

In [125]:
# the code below will find missing values in 'df_prods'
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [126]:
# The code below will create a subset containing the missing value from 'product_name' column
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [127]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [128]:
df_prods.shape

(49693, 5)

In [129]:
# the code below will create a new dataframe from 'df_prods' where the missing value is equal to false.
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [130]:
df_prods_clean.shape

(49677, 5)

In [131]:
# The code below will look for full duplicates in your 'df_prods' data
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [132]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [133]:
df_prods_clean.shape

(49677, 5)

In [134]:
# the code below will create a new dataframe that will include the duplicates that were identified above
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [135]:
df_prods_clean_no_dups.shape

(49672, 5)

In [136]:
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
5               6                                       Dry Nose Oil   
6               7                     Pure Coconut Water With Orange   
7               8                  Cut Russet Potatoes Steam N' Mash   
8               9                  Light Strawberry Blueberry Yogurt   
9              10     Sparkling Orange Juice & Prickly Pear Beverage   
10             11                                  Peach Mango Juice   
11             12                         Chocolate Fudge Layer Cake   
12             13                                  Saline Nasal Mist   
13             14                     Fresh Scent Dishwasher Cleaner   
14             15                           Overnight Diapers Size 6   
15             16                      Mint Chocolate Flavored Syrup   
16             17                                  Rendered Duck Fat   
17             18                Pizza for One Suprema  Frozen Pizza   
18             19   Gluten Free Quinoa Three Cheese & Mushroom Blend   
19             20     Pomegranate Cranberry & Aloe Vera Enrich Drink   
20             21                   Small & Medium Dental Dog Treats   
21             22                  Fresh Breath Oral Rinse Mild Mint   
22             23                             Organic Turkey Burgers   
23             24  Tri-Vi-Sol® Vitamins A-C-and D Supplement Drop...   
24             25            Salted Caramel Lean Protein & Fiber Bar   
25             26        Fancy Feast Trout Feast Flaked Wet Cat Food   
26             27  Complete Spring Water Foaming Antibacterial Ha...   
27             28                                  Wheat Chex Cereal   
28             29  Fresh Cut Golden Sweet No Salt Added Whole Ker...   
29             30         Three Cheese Ziti, Marinara with Meatballs   
...           ...                                                ...   
49663       49659                            Organic Creamed Coconut   
49664       49660                        Professionals Sleek Shampoo   
49665       49661                                              Porto   
49666       49662                       Bacon Cheddar Pretzel Pieces   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   
49668       49664                     Lemon Cayenne Drinking Vinegar   
49669       49665      Super Dark Coconut Ash & Banana Chocolate Bar   
49670       49666                      Ginger Snaps Snacking Cookies   
49671       49667           Enchilada with Spanish Rice & Beans Meal   
49672       49668                     Apple Cinnamon Scented Candles   
49673       49669                                   K Cup Dark Blend   
49674       49670                                Beef Summer Sausage   
49675       49671                               Milk Chocolate Drops   
49676       49672                             Cafe Mocha K-Cup Packs   
49677       49673              Stone Baked Multi Grain Artisan Rolls   
49678       49674            Frozen Greek Yogurt Bars Chocolate Chip   
49679       49675                Cinnamon Dolce Keurig Brewed K Cups   
49680       49676                             Ultra Red Energy Drink   
49681       49677                    Thick & Chunky Sloppy Joe Sauce   
49682       49678                    Large Chicken & Cheese Taquitos   
49683       49679                            Famous Chocolate Wafers   
49684       49680                 All Natural Creamy Caesar Dressing   
49685       49681            Spaghetti with Meatballs and Sauce Meal   
49686       49682                              

In [ ]:
# The code below show the steps for exporting the new cleaned 'df_prods'dataframe into a csv file

In [137]:
path_deps_clean = r'/Users/robertochidiac/Desktop/Instacart Basket Analysis/Data/Prepared Data'

In [138]:
df_prods_clean_no_dups.to_csv(os.path.join(path_deps_clean, 'products_checked.csv'))

# 04. Data Consistency Check for 'df_ords' DataFrame

In [139]:
# The cell below will use the 'df.describe()' function to investigate the 'df_ords' DataFrame
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

### After investigating the result of the function df.describe() I can't see any apparent issue with the data

In [ ]:
# Check for mixed-type data in your df_ords dataframe.

In [140]:
# the code below will check for mixed-type in the dataframe
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

### Running the function above should have given the column that had mixed-type data, but since it didn't come up with anything, it means that the dataframe has no mixed-type data

In [ ]:
#Run a check for missing values in your df_ords dataframe.
#In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [141]:
# the code below will find missing values in 'df_ords'
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### The result for the missing values function above came up with a result of 206209 missing values for column 'days_since_prior_order.
### My assumption for such missing value for that column could be that there is simply no information as to how many days went by since the last order from that customer

In [ ]:
# Address the missing values using an appropriate method.
# In a markdown cell, explain why you used your method of choice.

In [142]:
# First step is to write a code that will create a subset containing the missing value from 
#'days_since_prior_order' column
df_nan_ords = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [94]:
df_nan_ords

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
50               50   2086598        6    prior             1   
54               54   2565571        7    prior             1   
75               75    600894        8    prior             1   
79               79    280530        9    prior             1   
83               83   1224907       10    prior             1   
89               89   2411567       11    prior             1   
97               97   1483446       12    prior             1   
103             103   2618231       13    prior             1   
116             116   2906263       14    prior             1   
130             130   2514324       15    prior             1   
153             153   1273676       16    prior             1   
160             160   1737705       17    prior             1   
201             201   2780889       18    prior             1   
208             208   1321974       19    prior             1   
218             218    947722       20    prior             1   
223             223   2625565       21    prior             1   
257             257   2300537       22    prior             1   
273             273   1792521       23    prior             1   
278             278    611214       24    prior             1   
297             297   2610576       25    prior             1   
301             301   3229714       26    prior             1   
314             314    361493       27    prior             1   
396             396   1138968       28    prior             1   
421             421   2064090       29    prior             1   
440             440    375303       30    prior             1   
...             ...       ...      ...      ...           ...   
3420548     3420548    566068   206180    prior             1   
3420567     3420567    351920   206181    prior             1   
3420582     3420582    485448   206182    prior             1   
3420593     3420593     31169   206183    prior             1   
3420608     3420608    180109   206184    prior             1   
3420613     3420613   2168901   206185    prior             1   
3420624     3420624   1831589   206186    prior             1   
3420628     3420628    474057   206187    prior             1   
3420663     3420663   1944921   206188    prior             1   
3420671     3420671   2966736   206189    prior             1   
3420678     3420678   1955107   206190    prior             1   
3420687     3420687    936117   206191    prior             1   
3420693     3420693   3242108   206192    prior             1   
3420708     3420708    313609   206193    prior             1   
3420750     3420750   3395125   206194    prior             1   
3420762     3420762    926652   206195    prior             1   
3420782     3420782    825502   206196    prior             1   
3420787     3420787   3325493   206197    prior             1   
3420811     3420811   2154947   206198    prior             1   
3420819     3420819   3302466   206199    prior             1   
3420839     3420839   2508134   206200    prior             1   
3420863     3420863     28210   206201    prior             1   
3420896     3420896    799173   206202    prior             1   
3420919     3420919   1991850   206203    prior             1   
3420925     3420925   1438269   206204    prior             1   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    pr

In [143]:
# The code below will give us the number of columns and row for the 'df_ords' dataframe, which will give us 
# an overall view and comparaison with the new dataframe created above
df_ords.shape

(3421083, 8)

In [144]:
# With the median known, the new step would be to use the function in the code below to replace my missing
# values with the median
df_ords['ordered_today'] = np.where(df_ords['days_since_prior_order'].isnull(), True, False)

In [145]:
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           1   2398795        1    prior             2                   3   
2           2    473747        1    prior             3                   3   
3           3   2254736        1    prior             4                   4   
4           4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  ordered_today  
0                  8                     NaN           True  
1                  7                    15.0          False  
2                 12                    21.0          False  
3                  7                    29.0          False  
4                 15                    28.0          False

In [146]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
ordered_today                  0
dtype: int64

In [147]:
df_ords.shape

(3421083, 9)

### In this particular context just replacing the missing values with a specific median or mean would distort the data. So instead of replacing the NaN values we will create a new column called 'ordered_today that will give us a value of True when the column 'days_since_prior_order' has an NaN value and False for any other value in that column. 

In [ ]:
# Run a check for duplicate values in your df_ords data.
# In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [148]:
# The code below will look for full duplicates in 'df_prods' data
df_dups_ords = df_ords[df_ords.duplicated()]

In [149]:
df_dups_ords

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, ordered_today]
Index: []

### After runnining the duplicate function in my dataframe the result just the columns headers, which mean that the dataframe doesn't have any full duplicates. And if any duplicates were present the proper way to adress them would have been to remove them using 'df.drop_duplicates()' function

In [ ]:
# Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder 
# and give them appropriate, succinct names.

In [ ]:
# The code below show the steps for exporting the new cleaned 'df_prods' and df_ords dataframe into a csv file

In [150]:
path_deps_clean = r'/Users/robertochidiac/Desktop/Instacart Basket Analysis/Data/Prepared Data'

In [151]:
df_prods_clean_no_dups.to_csv(os.path.join(path_deps_clean, 'products_checked.csv'))

In [152]:
path_ords_clean = r'/Users/robertochidiac/Desktop/Instacart Basket Analysis/Data/Prepared Data'

In [154]:
df_ords.to_csv(os.path.join(path_ords_clean, 'orders_checked.csv'))